Migrate zendframework/zendframework package to component packages

weierophinney opened the pull request:
Currently, if you try and migrate a project that requires the "zendframework/zendframework" package, that package is not rewritten; it is instead retained verbatim. This patch modifies that behavior to instead replace the "zendframework/zendframework" package with the appropriate set of component packages based on the version constraint used.

To accomplish this, the patch adds two features:

- A concept of a "special case" for migrations. These are provided the filename and the file contents (`match(string $filename, string $contents): bool`) in order to determine if they should act; if so, they are then provided the contents (`replace(string $contents): string`) on which to perform replacements.

- A `ComposerJsonZendFrameworkPackageSpecialCase` that operates on `composer.json` files that refer to the "zendframework/zendframework" package.

This latter compares the version against a list of version ranges in order to deliver a list of packages to use instead (I performed iterative diffs between the `composer.json` contents of all zendframework/zendframework package versions to determine when changes occurred). When "self.version" is used to specify a dependency constraint (which happened in versions prior to 2.5.0), the class resolves to the latest version of the package that will fulfill the constraint (using the composer/semver package).

I have provided tests to validate that the correct replacements are made, and have tested with some local projects to verify that an installable set of dependencies are available when complete.  (Interestingly, I discovered that we were missing tags for versions 2.4.3 - 2.4.13 of zend-xmlrpc (and thus laminas-xmlrpc); I created and pushed these based on the 2.4.2 tag, as there were no changes in those versions.)

Once merged, this approach will allow users of the zendframework/zendframework package to properly migrate to Laminas finally.

weierophinney committed:
feat: Migrate zendframework/zendframework package to component packages

Currently, if you try and migrate a project that requires the
"zendframework/zendframework" package, that package is not rewritten; it
is instead retained verbatim.

This patch adds two things:

- A concept of a "special case" for migrations. These are provided the
  filename and the file contents in order to determine if they should
  act; if so, they are then provided the contents on which to perform
  replacements.

- A "ComposerJsonZendFrameworkPackageSpecialCase" that operates on
  `composer.json` files that refer to the "zendframework/zendframework"
  package.

This latter compares the version against a list of version ranges in
order to deliver a list of packages to use instead. When "self.version"
was used to determine which package version to use (generally on
versions less than 2.5.0), it resolves to the latest version of the
package that will fulfill the constraint.

I have provided tests to validate that the correct replacements are
made, and have tested with some local projects to verify that an
installable set of dependencies are available when complete.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,699 @@
+    public function matches($filename, $content)
+    {
+        return (bool) preg_match('#\bcomposer\.json$#', $filename)
+            && null !== strpos($content, '"zendframework/zendframework": "');
```
I would prefer here check if we have it set in require section, the above probably works for 99% but there might be very edge case where it will be wrong.

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,699 @@
+    {
+        $matches = [];
+        preg_match(
+            '#^(?P<line>(?P<indent>\s+)"zendframework/zendframework":\s*"(?P<constraint>[^"]+)"\s*(?P<eol>,\s*|$))#m',
```
The same here, I'd prefer operation on decoded json, instead of using regular expressions.
`composer.json` must be valid to preform update, so I think we should be fine.

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,699 @@
+    {
+        $semver   = new Semver();
+        $versions = array_keys(self::ZF_VERSIONS);
+        usort($versions, function ($a, $b) {
```
closure can be `static`

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,699 @@
+        'zendframework/zend-version'          => '~2.5.0',
+        'zendframework/zend-view'             => '~2.5.0',
+        'zendframework/zend-xmlrpc'           => '~2.5.0',
+        'zendframework/zendxml'               => '~1.0'
```
missing trailing comma, couple below as well

michalbundyra commented on test/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCaseTest.php:
```diff
@@ -0,0 +1,78 @@
+        $this->assertNotRegExp('#"zendframework/zendframework":#s', $replacement);
+
+        foreach ($expectedPackageList as $package => $constraint) {
+            $expected = sprintf('#"%s": "%s"#s', preg_quote($package), preg_quote($constraint));
```
Minor: Probably it is fine, but we are not testing here commas. It is quite easy to make a mistake, but we need a comma between each entry and the trailing comma is not allowed.
That's why I'd prefer in fixtures "result" file next to the source `composer.json`, so then we can just check if the result is equal to the result file.

michalbundyra reviewed (commented):
@weierophinney Great job! In general looks good, just some small minor comments. Thanks!

weierophinney committed:
fix: remove unused constants

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: better testing strategy

Per comments from @michalbundyra, I've refactored the tests to provide
an "output" `composer.json` file to test assertions against. This
ensures that the replacements are done in a way that keeps the JSON
integrity (e.g. no trailing commas; commas after every list item; etc.),
and which also clearly demonstrates what we are looking for when done.

Tests continue to pass, which allows me to move on to refactoring the
special case class itself.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: use JSON operations when performing zendframework replacements

- Verify that the entry exists in `require` or `require-dev`.
- Make replacements by parsing JSON, updating structure, and reformatting to JSON.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Thanks for the thorough review!

I've refactored to add fixture result files and perform assertions against them. Additionally, I've updated the special case class to use JSON operations, parsing the structure to ensure we only replace if found in the require or require-dev sections, and then replacing directly within the composer structure, re-serializing it to JSON when done. All tests continue to pass!

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,670 @@
+
+        $composer = json_decode($content, true);
+        return isset($composer['require']['zendframework/zendframework'])
+            || isset($composer['require-dev']['zendframework/zendframework']);
```
very minor: mixed case where ignored in composer v1, so technically we still can have `ZendFramework/ZendFramework` and it will be the same. But I see that changing it in the whole implementation is a bit pain, unless we "normalise" it first.

also, I cannot see any tests with `require-dev`.

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,670 @@
+{
+    public function matches($filename, $content)
+    {
+        if (! preg_match('#\bcomposer\.json$#', $filename)) {
```
minor: probably no need for regular expression, some string function will be sufficient here

weierophinney replied:
Determining that a string is at the last position of a haystack is actually _harder_ than using a regexp in this case. And PCRE is fast enough for this particular use-case.

weierophinney replied:
Additionally, this ensures we find it at a word boundary, which could mean any of "composer.json", "/composer.json", or "\\composer.json" would work. That's something we cannot do as easily with combinations of `substr` and `strrpos`. PCRE is definitely the correct approach here.

michalbundyra commented on src/SpecialCase/ComposerJsonZendFrameworkPackageSpecialCase.php:
```diff
@@ -0,0 +1,670 @@
+    {
+        $composer = json_decode($content, true);
+
+        $isProduction = isset($composer['require']['zendframework/zendframework']);
```
I'd suggest small refactor here:
```php
$section = isset($composer['require']['zendframework/zendframework'])
    ? 'require'
    : 'require-dev';

$packages = $this->getPackageList($composer[$section]['zendframework/zendframework']);
$composer = $this->updateComposer($composer, $packages, $section);
```

so then we just have one ternary operation.

michalbundyra reviewed (commented):
@weierophinney Changes you've made looks good to me! :)
Just couple minor comments.

I think we should include at least one test with require-dev section as it is a valid case in the code.

weierophinney committed:
qa: add test case for require-dev with zendframework package

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: ensure we also cover case variants of the package name

Composer 1.X allows packages to have mixed case, and internally
normalizes them. We need to do the same when checking to see if the
zendframework package is listed as a requirement.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra reviewed (approved):
🚢

weierophinney committed:
docs: adds CHANGELOG entry for #42

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
